source/slang-capture-replay/replay/decoder-helper.h

```diff
@@ -0,0 +1,108 @@
+#ifndef SLANG_CAPTURE_DECODER_HELPER_H
+#define SLANG_CAPTURE_DECODER_HELPER_H
+
+#include <stdint.h>
+#include "../../../slang.h"
+#include "../../../slang-com-helper.h"
+#include "../util/capture-format.h"
+
+namespace SlangCapture {
+
+    // This class is used to allocate memory for the type decoder
+    class DecoderAllocatorSingleton
+    {
+    public:
+        static DecoderAllocatorSingleton* getInstance();
+        void* allocate(size_t size);
+        ~DecoderAllocatorSingleton();
+    private:
+        DecoderAllocatorSingleton() = default;
+        std::vector<void*> m_allocations;
+    };
+
+    class DecoderBase
+    {
+    public:
+        virtual ~DecoderBase() = default;
+        void* allocate(size_t size) { return m_allocator->allocate(size); }
+    protected:
+        DecoderAllocatorSingleton* m_allocator = DecoderAllocatorSingleton::getInstance();
+    };
+
+    // We don't allow pointer type to be used as a template parameter
+    template <typename T, typename U = typename std::enable_if< !std::is_pointer<T>::value >::type>
+    class ValueDecoder : public DecoderBase
+    {
+    public:
+        T& getValue() { return m_value;}
+
+    protected:
+        T m_value {};
+    };
+
+    template <typename T, typename = typename std::enable_if< !std::is_pointer<T>::value >::type>
+    class StructDecoder : public ValueDecoder<T>
+    {
+    public:
+        using Super = ValueDecoder<T>;
+        size_t decode(const uint8_t* buffer, int64_t bufferSize);
+    };
+
+    // We only allow pointer type to be used as a template parameter
+    template <typename T, typename U = typename std::enable_if< std::is_pointer<T>::value >::type>
+    class PointerDecoder : public DecoderBase
+    {
+    public:
+        void setPointer(void* data) { m_pointer = static_cast<T>(data); }
+        T getPointer() const             { return m_pointer; }
+        void setPointerAddress(uint64_t address) { m_pointerAddress = address; }
+        void setDataSize(size_t size) { m_dataSize = size; }
+        size_t getDataSize() const { return m_dataSize; }
+    private:
+        T m_pointer {nullptr};
+        uint64_t m_pointerAddress = 0;
```

Comment:
This one is decoded from the binary stream, and it's used for tracking the allocation purpose, we don't want to cast this to any other type.
So 64bit integer is good enough for this variable.

Note that for this kind application, since the capture file could hand over platforms, e.g. captured on windows, and want to replay on Linux, I always try my best to use explicit non-os-dependent data type (e.g. uint{#bits}_t) to store to the file.